src/main/java/org/sciborgs1155/robot/vision/Vision.java

```diff
@@ -0,0 +1,157 @@
+      visionSim.addCamera(cameraSim, VisionConstants.ROBOT_TO_CAM);
+
+      cameraSim.enableDrawWireframe(true);
+    }
+  }
+
+  /**
+   * The latest estimated robot pose on the field from vision data. This may be empty. This should
+   * only be called once per loop.
+   *
+   * @return An {@link EstimatedRobotPose} with an estimated pose, estimate timestamp, and targets
+   *     used for estimation.
+   */
+  public Optional<EstimatedRobotPose> getEstimatedGlobalPose(
+      PhotonPoseEstimator poseEstimator, PhotonCamera camera) {
+    var visionEst = poseEstimator.update();
+
+    double latestTimestamp = camera.getLatestResult().getTimestampSeconds();
+    boolean newResult = Math.abs(latestTimestamp - lastEstTimestamp) > 1e-5;
+    if (Robot.isSimulation()) {
+      visionEst.ifPresentOrElse(
+          est ->
+              getSimDebugField()
+                  .getObject("VisionEstimation")
+                  .setPose(est.estimatedPose.toPose2d()),
+          () -> {
+            if (newResult) getSimDebugField().getObject("VisionEstimation").setPoses();
+          });
+    }
+    if (newResult) lastEstTimestamp = latestTimestamp;
+    return visionEst;
+  }
+
+  public EstimatedRobotPose[] getEstimatedGlobalPoses() {
+    return Stream.of(frontPhotonEstimator, rearPhotonEstimator)
+        .map(PhotonPoseEstimator::update)
+        .flatMap(Optional::stream)
+        .toArray(EstimatedRobotPose[]::new);
+  }
+
+  /**
+   * The standard deviations of the estimated pose from {@link #getEstimatedGlobalPose()}, for use
+   * with {@link edu.wpi.first.math.estimator.SwerveDrivePoseEstimator SwerveDrivePoseEstimator}.
+   * This should only be used when there are targets visible.
+   *
+   * @param estimatedPose The estimated pose to guess standard deviations for.
+   */
+  public Matrix<N3, N1> getEstimationStdDevs(
+      Pose2d estimatedPose, PhotonCamera camera, PhotonPoseEstimator estimator) {
+    var estStdDevs = VisionConstants.SINGLE_TAG_STD_DEVS;
+    var targets = camera.getLatestResult().getTargets();
+    int numTags = 0;
+    double avgDist = 0;
+    for (var tgt : targets) {
+      var tagPose = estimator.getFieldTags().getTagPose(tgt.getFiducialId());
+      if (tagPose.isEmpty()) continue;
+      numTags++;
+      avgDist +=
+          tagPose.get().toPose2d().getTranslation().getDistance(estimatedPose.getTranslation());
+    }
+    if (numTags == 0) return estStdDevs;
+    avgDist /= numTags;
+    // Decrease std devs if multiple targets are visible
+    if (numTags > 1) estStdDevs = VisionConstants.MULTIPLE_TAG_STD_DEVS;
+    // Increase std devs based on (average) distance
+    if (numTags == 1 && avgDist > 4)
+      estStdDevs = VecBuilder.fill(Double.MAX_VALUE, Double.MAX_VALUE, Double.MAX_VALUE);
+    else estStdDevs = estStdDevs.times(1 + (avgDist * avgDist / 30));
+
+    return estStdDevs;
+  }
+
+  // ----- Simulation
+
+  public void simulationPeriodic(Pose2d robotSimPose) {
```

Comment:
use `addPeriodic` in `Robot.java` to run this periodically when the robot is being simulated